File: d04/main.js
#!/usr/bin/env node

// Note: I am just learning JavaScript (not from "ground up", I know several languages like C, C++, Java, Python, ...),
// but more like how to work best with lists, dictionaires, etc., and getting "hands on experience"
// You will find a lot of boilerplate code here, and also some of my solutions might suffer from the fact that
// I am still learning. You have been warned... ;-)

import { readFileSync } from 'node:fs';
import { exit } from 'node:process';

function load_data(test_only) {
    if (test_only) {
        return readFileSync('./input_testdata.txt', 'utf8')
    } else {
        return readFileSync('./input.txt', 'utf8')
    }
}

if (process.argv.includes("-h") || process.argv.includes("--help")) {
    console.log("usage: node main.js [-h] [-s] [-1|2]")
    console.log("  -h:    this help")
    console.log("  -s:    execute large scale problem, not just example/test set, whiich is the default")
    console.log("  -1|2:  task 1 or 2 only, repectively (default: both tasks)")
    exit(1)
}

let tasks = process.argv.includes("-2") ? [2] : process.argv.includes("-1") ? [1] : [1, 2]
let test_only = process.argv.includes("-s") ? false : true

let start = performance.now()

// As part of my "boilerplate" I read data into an array of lines of strings
// and keep track of the dimension (m = number of columns, n = number of rows)

let data = load_data(test_only)
let lines = data.trim().split('\n')
let n = lines.length
let m = lines[0].length
console.log(`Size: ${n}x${m}`)

let load = performance.now()

// Actual problem solving starts here ===========================================================================

// Part 1 I originally solved in a somewhat stupid way, but leave it here for documentary purposes
// I just created an array of all relevant "lines":
//  - all existing lines
//  - all lines made up of existing columns
//  - all lines made up of existing diagonals (of varying length!)
// Then I used regexp to search for occurence of "XMAS" as well as "SAMX"

if (tasks.includes(1)) {
    let all_lines = []

    // all lines
    for (const l of lines) {
        all_lines.push(l)
    }
    
    // all columns
    for (let j = 0; j < m; j++) {
        let b = []
        for (let i = 0; i < n; i++) {
            b.push(lines[i][j])
        }
        all_lines.push(b.join(""))    
    }
    
    // all diagonals
    for (let x = m-4; x > 0; x--) {
        let b = []
        let cnt = 0
        while (x + cnt < m) {
            b.push(lines[cnt][x+cnt])
            cnt += 1
        }
        all_lines.push(b.join(""))
    }
    for (let x = 3; x < m; x++) {
        let b = []
        let cnt = 0
        while (x - cnt >= 0) {
            b.push(lines[cnt][x-cnt])
            cnt += 1
        }
        all_lines.push(b.join(""))    
    }
    for (let y = 0; y < n-3; y++) {
        let b = []
        let cnt = 0
        while (y + cnt < n) {
            b.push(lines[y+cnt][cnt])
            cnt += 1
        }
        all_lines.push(b.join(""))    
    }
    for (let y = 1; y < n-3; y++) {
        let b = []
        let cnt = 0
        while (y + cnt < n) {
            b.push(lines[y+cnt][m-1-cnt])
            cnt += 1
        }
        all_lines.push(b.join(""))    
    }
    // console.log(`total lines: ${all_lines.length}`)
    
    let cnt = 0
    for (const line of all_lines) {
        let xmas = line.match(/XMAS/g)
        let samx = line.match(/SAMX/g)  
        cnt += (xmas == null ? 0 : xmas.length)
        cnt += (samx == null ? 0 : samx.length)
    }
    console.log(`Part 1: ${cnt} occurences`)    
}

// Part 2 I solved by iterating over the grid, searching for 'A' and
// then checking if there is "M S" around it in all relevant directions

if (tasks.includes(2)) {
    let indices = []
    for (let i = 1; i < n-1; i++) {
        for (let j = 1; j < m; j++) {
            if (lines[i][j] == 'A') {
                let tl = lines[i-1][j-1]   // top-left
                let tr = lines[i-1][j+1]   // top-right
                let bl = lines[i+1][j-1]   // bottom-left
                let br = lines[i+1][j+1]   // bottom-right
                if ( (tl == 'M' && br == 'S' || tl == 'S' && br == 'M') &&
                     (bl == 'M' && tr == 'S' || bl == 'S' && tr == 'M')) {
                    indices.push([i, j])
                }
    
            }
        }
    }
    console.log(`Part 2: ${indices.length} occurences`)    
}

let stop = performance.now()

console.log(`Load took ${((load-start)/1000).toFixed(6)} msecs, calculation ${((stop-load)/1000).toFixed(6)} msecs`)

File: d04/main.py
#!/usr/bin/env python3

import re
from sys import argv
from time import process_time

tasks: list[int] = [1, 2]
test_only: bool = True

def solve():
    inputf = "input_testdata.txt" if test_only else "input.txt"

    start = process_time()

    puzzle = open(inputf).read()

    # idea: create 3 data structures:
    #  - 1: all rows appended
    #  - 2: all columns appended
    #  - 3: all diagonals appended
    # then search for "XMAS" and "SMAX" in each

    lines = puzzle.strip().splitlines()
    n = len(lines)
    m = len(lines[0])

    # 1
    one = "".join(lines)
    one_lines = lines

    # 2
    buff = []
    two_lines = []
    for j in range(m):
        b = []
        for i in range(n):
            buff.append(lines[i][j])
            b.append(lines[i][j])
        two_lines.append(''.join(b))
    two = ''.join(buff)

    # 3
    buff = []
    three_lines = []
    for x in range(m-4, 0, -1):
        b = []
        cnt = 0
        while x + cnt < m:
            b.append(lines[cnt][x+cnt])
            buff.append(lines[cnt][x+cnt])
            cnt += 1
        three_lines.append(''.join(b))
    for x in range(3, m):
        b = []
        cnt = 0
        while x - cnt >= 0:
            b.append(lines[cnt][x-cnt])
            buff.append(lines[cnt][x-cnt])
            cnt += 1
        three_lines.append(''.join(b))


    for y in range(n-3):
        b = []
        cnt = 0
        while y + cnt < n:
            b.append(lines[y+cnt][cnt])
            buff.append(lines[y+cnt][cnt])
            cnt += 1
        three_lines.append(''.join(b))
    for y in range(1, n-3):
        b = []
        cnt = 0
        while y + cnt < n:
            b.append(lines[y+cnt][m-1-cnt])
            buff.append(lines[y+cnt][m-1-cnt])
            cnt += 1
        three_lines.append(''.join(b))

    three = ''.join(buff)

    all_lines = one_lines + two_lines + three_lines
    load = process_time()

    # action
    if 1 in tasks:
        cnt = 0
        for l in all_lines:
            cnt += len(re.findall(r'XMAS', l))
            cnt += len(re.findall(r'SAMX', l))
        # cnt += len(re.findall(r'XMAS', two))
        # cnt += len(re.findall(r'SAMX', two))
        # cnt += len(re.findall(r'XMAS', three))
        # cnt += len(re.findall(r'SAMX', three))
        print(f"Task 1: {cnt} occurences")
        # task 1: ...
        pass
    
    if 2 in tasks:
        cnt = 0
        lst = []
        for i in range(1, m-1):
            for j in range (1, n-1):
                if (lines[i][j] == 'A'):
                    lt = lines[i-1][j-1]
                    rt = lines[i-1][j+1]
                    lb = lines[i+1][j-1]
                    rb = lines[i+1][j+1]
                    if (lt == 'M' and rb == 'S' or lt == 'S' and rb == 'M') and (rt == 'M' and lb == 'S' or rt == 'S' and lb == 'M'):
                        cnt += 1
                        lst.append((i,j))
        print(f"Task 2: {cnt} occurences")
        # task 2: ...
        pass

    stop = process_time()

    print(f"Load took {load-start:3.6f} secs, calculation {stop-load:3.6f} secs")


def main(argv):
    global tasks, test_only

    if "-h" in argv or "--help" in argv:
        print("usage: python3 main.py [-h] [-s] [-1|2]")
        print("  -h:    this help")
        print("  -s:    execute large scale problem, not just example/test set, which is the default")
        print("  -1|2:  task 1 or 2, respectively (default: both tasks)")
        exit(1)

    tasks = [2] if "-2" in argv else [1] if "-1" in argv else [1,2]

    test_only = False if "-s" in argv else True

    print(f"\n===> Running task(s) {tasks} in {'test' if test_only else 'full'} mode ===================================\n")

    solve()


if __name__ == "__main__":
    main(argv)

    # read, eg, 5,6, and return as list of int tupels
    # input = "5,6\n3,4\n2,3"
    # lst = list(tuple((int(x), int(y))) for x, y in re.findall('([0-9]+),([0-9]+)', input))

    # read alphaanumeric (incl '_'), and return as list of lists
    # input = "key -> value\nk2 -> val_2\nk3 -> v3"
    # lst = list( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print (lst)

    # ... or as dict
    # dct = dict( [ x, y ] for x, y in re.findall('([\w]+) -> ([\w]+)', input))
    # print(dct)

